### PR TITLE
feat(organisms): Burger Menu

### PR DESCRIPTION
**Pull request by gudwin**
## Changes
- New component Burger Menu added into bodiless-organisms
- header uses new design API
- Added Burger Menu to the site header (visible only on mobile screens)
- Added new page /burger-menu/ with burger-menu example
- Basic Documentation for the Burger Menu
- updated package* files with new dependency - negomi/react-burger-menu
- Submenus in Burger Menu implemented as accordions
- If menu item with submenu is a link, than "Overview" link added to accordion body. "Overview" text is configurable by site-builder
- Updates for Main Menu: new HOC's: asMainMenu, asEditableMainMenu
## Technical Details

- Extended the site tailwind.config
- Burger Menu API matches Main Menu API
- new HOC - asStatic sets the component and its children to read-only mode
- fix for List, preventing unwanted properties to appear into static mode
- Standardized API for applying responsive classes between
- withEditableTitle HOC: Allow injection of asEditable to support different kinds of editors

## Test Instructions

1. Add to site's Main Menu next items:
  - Item with link and with submenu
  - Item without link and with submenu
  - Item with link and without submenu
  - Item without link and without submenu
2. Resize the browser screen to small size
3. Check that burger menu button is visible
4. Open the burger menu
5. Check that close button works as expected
6. Check that items with submenu rendered as accordionns
7. Check that item with link and with submenu has an "Overview" link in the accordion body

## References:

- https://www.npmjs.com/package/react-burger-menu

### CODE REVIEW

File: packages/bodiless-organisms/src/components/MainMenu/asEditableMainMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import asMainMenu from './asMainMenu';
+import withEditableTitle from './withEditableTitle';
+
+/**
+ * HOC, produces *editable* main menu (based on rc-menu)
+ */
+const asEditableMainMenu = flow(
+  asEditableList,
+  asMainMenu,
+  withEditableTitle(asEditable),
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asEditableMainMenu.tsx]: This also should be moved to site level.

File: packages/bodiless-organisms/src/components/MainMenu/asEditableMenu.tsx [hunk 2]
```diff
@@ -23,7 +23,7 @@ import withEditableTitle from './withEditableTitle';
 const asEditableMenu = flow(
   asEditableList,
   asMenu,
-  withEditableTitle,
+  withEditableTitle(asEditable),
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asEditableMenu.tsx]: And this should be moved to site level.

File: packages/bodiless-organisms/src/components/MainMenu/asEditableSubMenu.tsx [hunk 3]
```diff
@@ -23,7 +23,7 @@ import withEditableTitle from './withEditableTitle';
 const asEditableSubMenu = flow(
   asEditableList,
   asSubMenu,
-  withEditableTitle,
+  withEditableTitle(asEditable),
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asEditableSubMenu.tsx]: And this should be moved to site level.

File: packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx [hunk 4]
```diff
@@ -19,14 +19,13 @@ import {
...
+import asMainMenu from './asMainMenu';
 /**
  * HOC, that incorporate the given component (usually based on <List /> component)
  * with rc-menu <Submenu /> component
  */
 const asSubMenu = flow(
-  asMenu,
+  asMainMenu,
   withDesign({
     Wrapper: replaceWith<any>(stylable(SubMenu)),
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx]: Why `<any>`

File: packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx [hunk 5]
```diff
@@ -19,14 +19,13 @@ import {
...
   replaceWith,
 } from '@bodiless/fclasses';
-import asMenu from './asMenu';
-
+import asMainMenu from './asMainMenu';
 /**
  * HOC, that incorporate the given component (usually based on <List /> component)
  * with rc-menu <Submenu /> component
  */
 const asSubMenu = flow(
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx]: Perhaps we should call this asMainSubMenu - or rename both to asRCMenu and asRCSubMenu?

File: packages/bodiless-organisms/src/components/MainMenu/withEditableTitle.tsx [hunk 6]
```diff
@@ -12,21 +12,22 @@
...
+import { withDesign, HOC } from '@bodiless/fclasses';
+import { asBodilessLink, withLinkToggle } from '@bodiless/components';
 import { withNode, withNodeKey } from '@bodiless/core';
 
+type Editable = (nodeKey?: string, placeholder?: string) => HOC;
 /**
  * HOC, that injects editable link toggle component into component design (usually Menu or Submenu)
  */
-const withEditableTitle = withDesign({
+const withEditableTitle = (editable: Editable) => withDesign({
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/withEditableTitle.tsx]: Can we rename this param: type  to `asEditable: AsEditable` -- I think it's clearer what it is.

File: examples/test-site/src/components/Menus/BurgerMenu/index.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+const HeaderContents: FunctionComponent = () => (
+  <A href="/">
+    <img src="/images/bodiless_logo.png" className="h-16" alt="Return To Home" />
+  </A>
+);
+
+const asBurgerMenuDefaultStyles = flow(
+  withDesign({
+    Wrapper: flow(
+      asMobileOnly,
```
[the remark this answers is not in the document]
Comment (gudwin) [reply examples/test-site/src/components/Menus/BurgerMenu/index.tsx]: Moved into `MainMenu.tsx`

File: examples/test-site/src/components/Layout/header.tsx [hunk 8]
```diff
@@ -53,14 +70,19 @@ const Header: FC<DesignableProps & { siteLogo: string }> = ({ design, siteLogo }
...
-  Container: flow(asPageContainer, addClasses('flex min-h-16 items-end')),
-  SiteReturn: flow(asHeader1, addClasses('flex-shrink')),
-});
+const asSiteHeader = flow(
+  designable(headerComponents),
+  withDesign({
+    Wrapper: flow(asPrimaryColorBackground, addClasses('')),
+    Container: flow(asPageContainer, addClasses('flex min-h-16 items-end')),
+    SiteReturn: flow(asHeader1, addClasses('flex-shrink')),
+    DesktopMenu: flow(asExceptMobile),
```
[the remark this answers is not in the document]
Comment (gudwin) [reply examples/test-site/src/components/Layout/header.tsx]: done

File: packages/bodiless-organisms/src/components/BurgerMenu/asEditableBurgerSubMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    Body: withDesign({
+      Header: flow(
+        withNodeKey('component'),
+        withLinkToggle,
+        withNodeKey('title'),
+      ),
+      InnerLinks: asEditableMenu,
+    }),
+  }),
+  withEditableTitle(asEditable),
```
[the remark this answers is not in the document]
Comment (gudwin) [reply packages/bodiless-organisms/src/components/BurgerMenu/asEditableBurgerSubMenu.tsx]: done

File: packages/bodiless-organisms/src/components/MainMenu/asEditableMainMenu.tsx (hunk 1 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asEditableMainMenu.tsx]: done

File: package.json [hunk 10]
```diff
@@ -98,6 +98,7 @@
     "npm-run-all": "^4.1.5",
     "ora": "^4.0.2",
     "prettier": "^1.17.0",
+    "react-burger-menu": "^2.6.13",
```
[the remark this answers is not in the document]
Comment (gudwin) [reply package.json]: removed

File: packages/bodiless-organisms/src/components/MainMenu/asEditableMenu.tsx (hunk 2 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asEditableMenu.tsx]: done

File: packages/bodiless-organisms/src/components/MainMenu/asEditableSubMenu.tsx (hunk 3 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asEditableSubMenu.tsx]: done

File: packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx (hunk 4 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx]: replaced with `ts-ignore` and added description

File: packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx (hunk 5 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asSubMenu.tsx]: renamed to asMainSubMenu, asEditableSubMenu also renamed too asEditableMainSubMenu

File: packages/bodiless-organisms/src/components/MainMenu/withEditableTitle.tsx (hunk 6 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/withEditableTitle.tsx]: Done

File: packages/bodiless-organisms/src/components/MainMenu/asMainSubMenu.tsx [hunk 11]
```diff
@@ -24,11 +24,12 @@ import asMainMenu from './asMainMenu';
  * HOC, that incorporate the given component (usually based on <List /> component)
  * with rc-menu <Submenu /> component
  */
-const asSubMenu = flow(
+const asMainSubMenu = flow(
   asMainMenu,
   withDesign({
-    Wrapper: replaceWith<any>(stylable(SubMenu)),
+    // @ts-ignore: Types of parameters are incompatible.
```
Comment (wodenx) [inline packages/bodiless-organisms/src/components/MainMenu/asMainSubMenu.tsx]: This is not sufficient documentation.  Please explain why it's not practical to resolve this type error.

**Comment by konst8** [comment]
Works as described in test instructions. Passed at a7892a410702b147ecbe7ebe29eb048afcf76b82.

File: packages/bodiless-organisms/src/components/MainMenu/asMainSubMenu.tsx (hunk 11 above)
Comment (gudwin) [reply packages/bodiless-organisms/src/components/MainMenu/asMainSubMenu.tsx]: Updated with an attempt to resolve a type error

**Comment by asiarheyeu** [comment]
According to requirements the breakpoints width of the screen when the main menu should turn into burger menu should be 1024px. Currently it toggles at 768px.

**STR**
Open the site at 1023 screen width.

**ER**
Burger menu is shown

**AR**
Main menu is shown

**Pull request merged**